include/nbl/video/ILogicalDevice.h

```diff
@@ -539,19 +543,133 @@ class ILogicalDevice : public core::IReferenceCounted
 
     virtual void waitIdle() = 0;
 
+    virtual void* mapMemory(const IDriverMemoryAllocation::MappedMemoryRange& memory, IDriverMemoryAllocation::E_MAPPING_CPU_ACCESS_FLAG accessHint = IDriverMemoryAllocation::EMCAF_READ_AND_WRITE) = 0;
+
+    virtual void unmapMemory(IDriverMemoryAllocation* memory) = 0;
+
+    inline void updateBufferRangeViaStagingBuffer(IGPUCommandBuffer* cmdbuf, IGPUFence* fence, IGPUQueue* queue, const asset::SBufferRange<IGPUBuffer>& bufferRange, const void* data)
+    {
+        auto* cmdpool = cmdbuf->getPool();
+        assert(cmdpool->getCreationFlags() & IGPUCommandPool::ECF_RESET_COMMAND_BUFFER_BIT);
+        assert(cmdpool->getQueueFamilyIndex() == queue->getFamilyIndex());
+
+        //EventHandle event = null;
+        resetFences(1u, &fence);
+        for (size_t uploadedSize=0ull; uploadedSize<bufferRange.size;)
+        {
+            const void* dataPtr = reinterpret_cast<const uint8_t*>(data) + uploadedSize;
+            uint32_t localOffset = video::StreamingTransientDataBufferMT<>::invalid_address;
+            uint32_t alignment = 64u; // smallest mapping alignment capability
+            uint32_t subSize = static_cast<uint32_t>(core::min<uint32_t>(core::alignDown(m_defaultUploadBuffer.get()->max_size(), alignment), bufferRange.size - uploadedSize));
+            m_defaultUploadBuffer.get()->multi_place(std::chrono::high_resolution_clock::now() + std::chrono::microseconds(500u), 1u, (const void* const*)&dataPtr, &localOffset, &subSize, &alignment);
+
+            // keep trying again
+            if (localOffset == video::StreamingTransientDataBufferMT<>::invalid_address)
+            {
+                // but first sumbit the already buffered up copies
+                cmdbuf->end();
+                IGPUQueue::SSubmitInfo submit;
+                submit.commandBufferCount = 1u;
+                submit.commandBuffers = &cmdbuf;
+                submit.signalSemaphoreCount = 0u;
+                submit.pSignalSemaphores = nullptr;
+                submit.waitSemaphoreCount = 0u;
+                submit.pWaitSemaphores = nullptr;
+                submit.pWaitDstStageMask = nullptr;
+                queue->submit(1u, &submit, fence); // <threadsafeSubmit==true> means submit while locking an internal mutex
+                waitForFences(1u, &fence, false, 9999999999ull);
+                //event.manuallySignal(); // ???????? just to be exact, probably would still work without, simply because we're resetting a fence that was previously queued up
```

Comment:
your event handler needs to "poll" the event which owns the fence you're about to reset, otherwise you'll miss the fact that its done